Mathlib/LinearAlgebra/Dimension.lean

```diff
@@ -0,0 +1,1449 @@
+    rcases Cardinal.lift_mk_le'.mp w₁ with ⟨f⟩
+    haveI : Infinite ι' := Infinite.of_injective f f.2
+    have w₂ := infinite_basis_le_maximal_linearIndependent' v' _ v.linearIndependent v.maximal
+    exact le_antisymm w₁ w₂
+#align mk_eq_mk_of_basis mk_eq_mk_of_basis
+
+/-- Given two bases indexed by `ι` and `ι'` of an `R`-module, where `R` satisfies the invariant
+basis number property, an equiv `ι ≃ ι' `. -/
+def Basis.indexEquiv (v : Basis ι R M) (v' : Basis ι' R M) : ι ≃ ι' :=
+  (Cardinal.lift_mk_eq'.1 <| mk_eq_mk_of_basis v v').some
+#align basis.index_equiv Basis.indexEquiv
+
+theorem mk_eq_mk_of_basis' {ι' : Type w} (v : Basis ι R M) (v' : Basis ι' R M) : (#ι) = (#ι') :=
+  Cardinal.lift_inj.1 <| mk_eq_mk_of_basis v v'
+#align mk_eq_mk_of_basis' mk_eq_mk_of_basis'
+
+end InvariantBasisNumber
+
+section RankCondition
+
+variable {R : Type u} [Ring R] [RankCondition R]
+
+variable {M : Type v} [AddCommGroup M] [Module R M]
+
+/-- An auxiliary lemma for `Basis.le_span`.
+
+If `R` satisfies the rank condition,
+then for any finite basis `b : Basis ι R M`,
+and any finite spanning set `w : Set M`,
+the cardinality of `ι` is bounded by the cardinality of `w`.
+-/
+theorem Basis.le_span'' {ι : Type _} [Fintype ι] (b : Basis ι R M) {w : Set M} [Fintype w]
+    (s : span R w = ⊤) : Fintype.card ι ≤ Fintype.card w := by
+  -- We construct an surjective linear map `(w → R) →ₗ[R] (ι → R)`,
+  -- by expressing a linear combination in `w` as a linear combination in `ι`.
+  fapply card_le_of_surjective' R
+  · exact b.repr.toLinearMap.comp (Finsupp.total w M R (↑))
+  · apply Surjective.comp (g := b.repr.toLinearMap)
+    apply LinearEquiv.surjective
+    rw [← LinearMap.range_eq_top, Finsupp.range_total]
+    simpa using s
+#align basis.le_span'' Basis.le_span''
+
+/--
+Another auxiliary lemma for `Basis.le_span`, which does not require assuming the basis is finite,
+but still assumes we have a finite spanning set.
+-/
+theorem basis_le_span' {ι : Type _} (b : Basis ι R M) {w : Set M} [Fintype w] (s : span R w = ⊤) :
+    (#ι) ≤ Fintype.card w := by
+  haveI := nontrivial_of_invariantBasisNumber R
+  haveI := basisFintypeOfFiniteSpans w s b
+  rw [Cardinal.mk_fintype ι]
+  simp only [Cardinal.natCast_le]
+  exact Basis.le_span'' b s
+#align basis_le_span' basis_le_span'
+
+-- Note that if `R` satisfies the strong rank condition,
+-- this also follows from `linearIndependent_le_span` below.
+/-- If `R` satisfies the rank condition,
+then the cardinality of any basis is bounded by the cardinality of any spanning set.
+-/
+theorem Basis.le_span {J : Set M} (v : Basis ι R M) (hJ : span R J = ⊤) : (#range v) ≤ (#J) := by
+  haveI := nontrivial_of_invariantBasisNumber R
+  cases fintypeOrInfinite J
+  · rw [← Cardinal.lift_le, Cardinal.mk_range_eq_of_injective v.injective, Cardinal.mk_fintype J]
+    convert Cardinal.lift_le.{w, v}.2 (basis_le_span' v hJ)
+    simp
+  · -- Porting note: `have` in this line is unnecessary.
```

Comment:
I'm happy if small emendations like this go through without a porting note. We don't need someone to review this after the port is complete.